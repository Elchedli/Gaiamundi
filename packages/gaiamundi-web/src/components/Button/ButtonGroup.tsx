import classNames from 'classnames';
import {
  Children,
  cloneElement,
  FC,
  PropsWithChildren,
  ReactElement,
  useMemo,
} from 'react';

import { ButtonComponentProps } from './Button';

export type ButtonGroupProps = PropsWithChildren<{
  pill?: boolean;
  outline?: boolean;
  className?: string;
}>;

const ButtonGroup: FC<ButtonGroupProps> = ({
  children,
  pill,
  outline,
  className,
}) => {
  const items = useMemo(
    () =>
      Children.map(
        children as ReactElement<ButtonComponentProps>[],
        (child, index) =>
          cloneElement(child, {
            pill,
            outline,
            positionInGroup:
              index === 0
                ? 'start'
                : index ===
                  (children as ReactElement<ButtonComponentProps>[]).length - 1
                ? 'end'
                : 'middle',
          })
      ),
    [children, outline, pill]
  );

  return (
    <div className={classNames('inline-flex', className)} role="group">
      {items}
    </div>
  );
};

ButtonGroup.displayName = 'Button.Group';
export default ButtonGroup;
