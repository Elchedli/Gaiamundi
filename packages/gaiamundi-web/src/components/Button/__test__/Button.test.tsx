import { fireEvent, render } from '@testing-library/react';

import { Button } from 'components/Button/Button';

describe('Button Component', () => {
  it('renders without crashing', () => {
    const { container } = render(<Button label="Button" />);
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('renders label', () => {
    const label = 'Button Label';
    const { container, getByText } = render(<Button label={label} />);
    const button = getByText(label);
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('handles click event', () => {
    const label = 'Button Label';
    const handleClick = jest.fn();
    const { container, getByText } = render(
      <Button label={label} onClick={handleClick} />
    );
    const button = getByText(label);
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it('renders color correctly', () => {
    const color = 'blue';
    const { container } = render(<Button label="Button" color={color} />);
    const button = container.firstChild;
    expect(button).toHaveClass(
      `text-white bg-blue-700 border border-transparent`
    );
    expect(container).toMatchSnapshot();
  });

  it('renders gradient monochrome correctly', () => {
    const gradientMonochrome = 'teal';
    const { container } = render(
      <Button label="Button" gradientMonochrome={gradientMonochrome} />
    );
    const button = container.firstChild;
    expect(button).toHaveClass(`from-teal-400`);
    expect(container).toMatchSnapshot();
  });

  it('renders gradient duo tone correctly', () => {
    const gradientDuoTone = 'cyanToBlue';
    const { container } = render(
      <Button label="Button" gradientDuoTone={gradientDuoTone} />
    );
    const button = container.firstChild;
    expect(button).toHaveClass(`from-cyan-500`);
    expect(container).toMatchSnapshot();
  });

  it('renders pill correctly', () => {
    const { container } = render(<Button label="Button" pill />);
    const button = container.firstChild;
    expect(button).toHaveClass(`rounded-full`);
    expect(container).toMatchSnapshot();
  });
  it('renders size correctly', () => {
    const { container } = render(<Button label="Button" size="lg" />);
    const button = container.firstChild;
    expect(button).toHaveTextContent('Button');
    expect(container).toMatchSnapshot();
  });

  it('calls onClick when clicked', () => {
    const onClick = jest.fn();
    const { container } = render(<Button label="Button" onClick={onClick} />);
    const button = container.firstChild;
    fireEvent.click(button!);
    expect(onClick).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
