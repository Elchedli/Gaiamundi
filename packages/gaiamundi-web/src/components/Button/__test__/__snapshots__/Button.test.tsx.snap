// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component calls onClick when clicked 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component handles click event 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button Label
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders color correctly 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders gradient duo tone correctly 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-gradient-to-r from-cyan-500 to-blue-500 hover:bg-gradient-to-bl focus:ring-4 focus:ring-cyan-300 dark:focus:ring-cyan-800"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders gradient monochrome correctly 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-gradient-to-r from-teal-400 via-teal-500 to-teal-600 hover:bg-gradient-to-br focus:ring-4 focus:ring-teal-300 dark:focus:ring-teal-800"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders label 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button Label
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders pill correctly 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-full text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders size correctly 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-base px-5 py-2.5 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;

exports[`Button Component renders without crashing 1`] = `
<div>
  <button
    class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
    data-testid="button-element"
    type="button"
  >
    <span
      class="flex items-center text-sm px-4 py-2 rounded-md"
    >
      <span
        class="ml-4 inline-flex h-4 w-4 items-center justify-center rounded-full bg-blue-200 text-xs font-semibold text-blue-800"
      >
        Button
      </span>
    </span>
  </button>
</div>
`;
